tests/desktop.nix

```diff
@@ -0,0 +1,114 @@
+#
+# SPDX-License-Identifier: MPL-2.0
+
+# Test that a desktop machine can boot
+
+(import ./lib.nix) {
+  name = "desktop-test";
+  nodes = {
+    desktop = { self, inputs, pkgs, outputs, ... }: {
+      imports =
+        [
+          self.nixosModules.desktopSystem
+          # TODO: update to user changes later
+          ../nixos/users/andrew
+          {
+            home-manager.extraSpecialArgs = { inherit inputs outputs; };
+            home-manager.users.andrew = self.homeManagerModules.andrew;
+          }
+        ];
+
+      ahayzen = {
+        testing = true;
+
+        hostName = "desktop";
+      };
+
+      environment.etc."ahayzen.com/expected/gnome".source = ./expected/gnome;
+
+      # A test user to login
+      users.users.test = {
+         description = "Test";
+         isNormalUser = true;
+         extraGroups = [ "wheel" ];
+         password = "test";
+      };
+
+      virtualisation = {
+        cores = 2;
+        diskSize = 4 * 1024;
+        memorySize = 2 * 1024;
+      };
+    };
+  };
+
+  extraPythonPackages = p: [ p.numpy p.opencv4 ];
+
+  testScript = ''
+    import cv2
+    import numpy as np
+    import os
+
+    start_all()
+
+    # From https://stackoverflow.com/a/51289984
+    def compare_images(filename):
+      path1 = os.path.join(desktop.out_dir, f"{filename}.png")
+
+      desktopPath2 = f"/etc/ahayzen.com/expected/gnome/{filename}.png"
+      desktop.copy_from_vm(desktopPath2, "expected")
+      path2 = os.path.join(desktop.out_dir, "expected", f"{filename}.png")
+
+      # Load the images
+      img1 = cv2.imread(path1, 0)
+      img2 = cv2.imread(path2, 0)
+
+      # Absolute difference
+      res = cv2.absdiff(img1, img2)
+
+      # Convert to an integer type
+      res = res.astype(np.uint8)
+
+      # Determine percetange difference
+      return (np.count_nonzero(res) * 100)/ res.size
+
+    #
+    # Test that the desktop machine boots
+    #
+
+    with subtest("Ensure that the system booted"):
+      desktop.send_key("ctrl-alt-f7")
+      desktop.wait_for_unit("default.target", timeout=120)
+
+    with subtest("Ensure that GDM is correct"):
+      desktop.sleep(1)
+
+      # Compare the screenshot
+      desktop.screenshot("login")
+      diff = compare_images("login")
+      assert diff < 0.05, f"Login image was different by {diff * 100}%, more than 5%"
+
+    with subtest("Ensure that we can login"):
+      desktop.send_chars("test\n", delay=0.2)
+      desktop.sleep(3)
+      desktop.send_chars("test\n", delay=0.2)
+
+    with subtest("Ensure that desktop is reached"):
+      # Wait long enough for GNOME Shell to login
+      #
+      # TODO: can we wait for something in the journal? like initial setup to start?
+      desktop.sleep(30)
+
+      # Compare the screenshot
+      desktop.screenshot("desktop")
+      diff = compare_images("desktop")
+      assert diff < 0.05, f"Desktop image was different by {diff * 100}%, more than 5%"
+
```

Comment:
can we do further tests such as using a real home manager setup and launch an application ?

See https://wiki.nixos.org/wiki/NixOS_VM_tests#wayland_application_example